plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -848,6 +857,280 @@ compileExpr e = withContextM 2 (sdToTxt $ "Compiling expr:" GHC.<+> GHC.ppr e) $
+                    | ty `GHC.eqType` GHC.integerTy ->
+                        Just . compileExpr . GHC.Lit $ GHC.LitNumber numTy (i * j)
+                _ -> Nothing
+            )
+          , ( 'GHC.Num.fromInteger
+            , \case
+                -- Compile `fromInteger @Integer` via `id`.
+                [GHC.Type ty, _numDict]
+                    | ty `GHC.eqType` GHC.integerTy -> Just $ do
+                        idId <- lookupId =<< thNameToGhcNameOrFail 'id
+                        compileExpr $ GHC.mkCoreApps (GHC.Var idId) [GHC.Type GHC.integerTy]
+                [GHC.Type ty, _numDict, arg]
+                    | ty `GHC.eqType` GHC.integerTy -> Just $ compileExpr arg
+                _ -> Nothing
+            )
+          , ( 'GHC.Num.negate
+            , \case
+                -- Compile `negate @Integer` via `subtractInteger 0`.
+                [GHC.Type ty, _numDict]
+                    | ty `GHC.eqType` GHC.integerTy -> Just $ do
+                        subtractIntegerId <- lookupId =<< thNameToGhcNameOrFail 'BI.subtractInteger
+                        compileExpr $
+                            GHC.mkCoreApps
+                                (GHC.Var subtractIntegerId)
+                                [GHC.Lit $ GHC.LitNumber GHC.LitNumInteger 0]
+                [GHC.Type ty, _numDict, GHC.Lit (GHC.LitNumber numTy i)]
+                    | ty `GHC.eqType` GHC.integerTy ->
+                        Just . compileExpr . GHC.Lit $ GHC.LitNumber numTy (-i)
+                _ -> Nothing
+            )
+          , ( 'GHC.Real.toInteger
+            , \case
+                -- Compile `toInteger @Integer` via `id`.
+                [GHC.Type ty, _integralDict]
+                    | ty `GHC.eqType` GHC.integerTy -> Just $ do
+                        idId <- lookupId =<< thNameToGhcNameOrFail 'id
+                        compileExpr $ GHC.mkCoreApps (GHC.Var idId) [GHC.Type GHC.integerTy]
+                [GHC.Type ty, _numDict, arg]
+                    | ty `GHC.eqType` GHC.integerTy -> Just $ compileExpr arg
+                _ -> Nothing
+            )
+          ]
+
+-- | Like `isKnownApp` but for class dictionaries.
+isKnownPred ::
+    CompilingDefault uni fun m ann =>
+    GHC.Type ->
+    GHC.Var ->
+    [GHC.CoreExpr] ->
+    Maybe (m (PIRTerm uni fun))
+isKnownPred ty v@(GHC.idDetails -> GHC.DataConWorkId _) args =
+    if GHC.isPredTy ty
+        then do
+            (tc, targs) <- GHC.splitTyConApp_maybe ty
+            cls <- GHC.tyConClass_maybe tc
+            f <- Map.lookup (splitGhcName (GHC.tyConName tc)) knownPreds
+            f targs (cls, v, args)
+        else Nothing
+isKnownPred _ _ _ = Nothing
+
+-- | Like `knownApps` but for class dictionaries.
+knownPreds ::
+    CompilingDefault uni fun m ann =>
+    Map (Maybe String, String)
+        ([GHC.Type] -> (GHC.Class, GHC.Var, [GHC.CoreExpr]) -> Maybe (m (PIRTerm uni fun)))
+knownPreds =
```

Comment:
Hm. This feels very complicated. I don't suppose there's any way that we can match the ids of the method functions themselves? That would make things a _ton_ easier: we would just substitute exactly those for our functions and we'd be done, i.e.

```
MkEqDict @Integer integerEq integerNEq -- spot integerEq and replace _there_
```